concrete-csprng/src/generators/aes_ctr/states.rs

```diff
@@ -0,0 +1,159 @@
+use crate::generators::aes_ctr::index::{AesIndex, TableIndex};
+
+/// A pointer to the next byte to be outputted by the generator.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub struct BufferPointer(pub usize);
+
+/// A structure representing the current state of generator using batched aes-ctr approach.
+#[derive(Debug, Clone, Copy)]
+pub struct State {
+    table_index: TableIndex,
+    buffer_pointer: BufferPointer,
+}
+
+/// A structure representing the action to be taken by the generator after shifting its state.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ShiftAction {
+    /// Yield the byte pointed to by the 0-th field.
+    YieldByte(BufferPointer),
+    /// Refresh the buffer starting from the 0-th field, and yield the byte pointed to by the 0-th
+    /// field.
+    RefreshBatchAndYieldByte(AesIndex, BufferPointer),
+}
+
+impl State {
+    /// Creates a new state from the initial table index.
+    pub fn new(table_index: TableIndex) -> Self {
+        State {
+            table_index: table_index.decremented(),
+            buffer_pointer: BufferPointer(127),
```

Comment:
I don't think that would make sense, because it would essentially make `State` contain most of what a generator is. I think I prefer to keep separated. Though, you are completely right that a constant would be better for this values. I'll change that everywhere.

Reply:
yeah I figured it would mean moving some stuff if we don't use a constant!

Reply:
I changed the code to use constants everywhere ! Thanks !